rust/agents/relayer/src/msg/processor.rs

```diff
@@ -198,3 +201,242 @@ impl MessageProcessorMetrics {
+        destination_domain: &HyperlaneDomain,
+        db: &HyperlaneRocksDB,
+        num_operations: usize,
+    ) -> Vec<Box<DynPendingOperation>> {
+        let (message_processor, mut receive_channel) =
+            dummy_message_processor(&origin_domain, &destination_domain, &db);
+
+        let process_fut = message_processor.spawn();
+        let mut pending_messages = vec![];
+        let pending_message_accumulator = async {
+            while let Some(pm) = receive_channel.recv().await {
+                pending_messages.push(pm);
+                if pending_messages.len() == num_operations {
+                    break;
+                }
+            }
+        };
+        tokio::select! {
+            _ = process_fut => {},
+            _ = pending_message_accumulator => {},
+            _ = sleep(Duration::from_millis(200)) => { panic!("No PendingMessage received from the processor") }
+        };
+        pending_messages
+    }
+
+    #[tokio::test]
+    async fn test_full_pending_message_persistence_flow() {
+        test_utils::run_test_db(|db| async move {
+            let origin_domain = dummy_domain(0, "dummy_origin_domain");
+            let destination_domain = dummy_domain(1, "dummy_destination_domain");
+            let db = HyperlaneRocksDB::new(&origin_domain, db);
+
+            // Assume the message syncer stored some new messages in HyperlaneDB
+            let msg_retries = vec![0, 0, 0];
+            persist_retried_messages(&msg_retries, &db, &destination_domain);
+
+            // Run parser to load the messages in memory
+            let pending_messages = get_first_n_operations_from_processor(
+                &origin_domain,
+                &destination_domain,
+                &db,
+                msg_retries.len(),
+            )
+            .await;
+
+            // Set some retry counts. This should update HyperlaneDB entries too.
+            let msg_retries_to_set = vec![3, 0, 10];
+            pending_messages
+                .into_iter()
+                .enumerate()
+                .for_each(|(i, mut pm)| pm.set_retries(msg_retries_to_set[i]));
+
+            // Run parser again
+            let pending_messages = get_first_n_operations_from_processor(
+                &origin_domain,
+                &destination_domain,
+                &db,
+                msg_retries.len(),
+            )
+            .await;
+
+            // Expect the HyperlaneDB entry to have been updated, so the `OpQueue` in the submitter
+            // can be accurately reconstructed on restart.
+            // If the retry counts were correctly persisted, the backoffs will have the expected value.
+            pending_messages
+                .iter()
+                .zip(msg_retries_to_set.iter())
+                .for_each(|(pm, expected_retries)| {
+                    // Round up the actuall backoff because it was calculated with an `Instant::now()` that was a fraction of a second ago
```

Comment:
hope this doesn't prove to add some nondeterminism in the unit tests? Probably is fine but if it ever does provide issues we can consider something like https://docs.rs/mock_instant/latest/mock_instant/

Reply:
On my local machine, the difference between instants was less than 0.01s, so I think it's very unlikely that it would ever exceed 0.5s and be rounded the wrong way. Let's see, and I'll keep `mock_instant` in mind 👍